### PR TITLE
Rework configuration and create generic interface for new oracles

### PR DESCRIPTION
**Pull request by oeble**
Implements #24 #31 #32

### CODE REVIEW

File: programs/scope/src/utils/ctokens.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+use solend_program::state::Reserve;
+
+const DECIMALS: u32 = 15u32;
+const FACTOR: u64 = 10u64.pow(DECIMALS);
+
+pub fn get_price(solend_reserve_account: &AccountInfo) -> Result<DatedPrice> {
+    let reserve = Reserve::unpack(&solend_reserve_account.data.borrow())?;
+    let rate = reserve.collateral_exchange_rate()?;
+
+    let value = rate.liquidity_to_collateral(FACTOR)?;
```
Comment (gmmour1) [inline programs/scope/src/utils/ctokens.rs]: is this an inbuilt conversion method?

File: programs/scope/src/utils/mod.rs [hunk 2]
```diff
@@ -21,18 +22,50 @@ pub fn check_context<T>(ctx: &Context<T>) -> ProgramResult {
...
+/// This function shall be called before update of oracle mappings
+pub fn validate_oracle_account(
+    price_type: OracleType,
+    price_account: &AccountInfo,
+) -> crate::Result<()> {
+    match price_type {
+        OracleType::Pyth => pyth::validate_pyth_price_info(price_account),
+        OracleType::SwitchboardV1 => Ok(()), // TODO at least check account ownership?
+        OracleType::SwitchboardV2 => Ok(()), // TODO at least check account ownership?
+        OracleType::YiToken => Ok(()),       // TODO how shall we validate yi token account?
```
Comment (gmmour1) [inline programs/scope/src/utils/mod.rs:68]: these are basically static accounts, so we can go back to validating against some hardcoded (in config) pub keys for main net, devnet etc. Conversely, this is also checked by the way your config works, but essentially for this token type there can only be one correct account per solana env

File: programs/scope/src/utils/yitoken.rs [hunk 3]
```diff
@@ -3,26 +3,79 @@ use anchor_lang::prelude::*;
...
+
+    // fees in millibps
+    pub stake_fee: u32,
+    pub unstake_fee: u32,
+}
+
+/// Compute the current price
+///
+/// Return `None` in case of overflow
+pub fn price_compute(tokens_amount: u64, mint_supply: u64) -> Option<Price> {
```
Comment (gmmour1) [inline programs/scope/src/utils/yitoken.rs:30]: I wonder if we should include the stake and unstake fees in the computation the same as their source code, even though they assured us that they're not gonna change them. We discussed this before and @y2kappa said we were ok then. But now that you've imported the whole yi account and you're deserialising and getting the stake and unstake fees, maybe it makes sense to completely replicate their logic from their source code, it's a few extra lines of code: https://github.com/CrateProtocol/yi/blob/master/programs/yi/src/state.rs

    pub fn calculate_underlying_for_yitokens(
        &self,
        yitoken_amount: u64,
        total_underlying_tokens: u64,
        total_supply: u64,
    ) -> Option<u64> {
        if yitoken_amount == 0 {
            return Some(0);
        }
        // impossible to have more yitokens than the total supply.
        if yitoken_amount > total_supply {
            return None;
        }
        // if withdrawing all tokens, give the entire supply
        if yitoken_amount == total_supply {
            return Some(total_underlying_tokens);
        }
        let amt_no_fee = (yitoken_amount as u128)
            .checked_mul(total_underlying_tokens.into())?
            .checked_div(total_supply.into())?
            .to_u64()?;
        if self.unstake_fee_millibps == 0 {
            Some(amt_no_fee)
        } else {
            (amt_no_fee as u128)
                .checked_mul(self.unstake_fee_millibps.into())?
                .checked_div(MILLIBPS_PER_WHOLE.into())?
                .to_u64()
        }
    }

File: tests/utils.ts [hunk 4]
```diff
@@ -16,39 +21,141 @@ export enum Tokens {
...
+  CBTC,
+  CMSOL,
 }
 
-export enum PriceType {
-  Pyth = 0,
-  SwitchboardV1 = 1,
-  SwitchboardV2 = 2,
-  YiToken = 3,
+export const initialTokens: ITokenInput[] = [
```
Comment (gmmour1) [inline tests/utils.ts:30]: As discussed with @oeble one idea here would be to store the price as mantissa and exponent, mainly to catch the cToken edge case with the 15 significant digits that won't work with JS/TS Decimal. Also, we store prices as mantissa and expo in Rust anyway, so it is pretty consistent. This is a small detail so we can look into it in the future as well.

**Review by gmmour1 — COMMENTED** [review]
Amazing PR, only minor stylistic comments and clarifying questions.

File: off_chain/scope-cli/src/oracle_helpers/single_account_oracle.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SingleAccountOracle")
+            .field("label", &self.label)
+            .field("oracle_account", &self.oracle_account)
+            .field("oracle_type", &self.oracle_type)
+            .finish()
+    }
+}
+
+impl TokenEntry for SingleAccountOracle {}
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/oracle_helpers/single_account_oracle.rs:76]: Yes, you cannot impl it without the others, but you need to explicitly implement it.

File: off_chain/scope-cli/src/scope_client.rs [hunk 6]
```diff
@@ -141,254 +125,215 @@ impl ScopeClient {
...
-                    )
-                })
+            .zip(onchain_types)
+            .filter(|((_, &oracle_mapping), _)| oracle_mapping != zero_pk)
+            .map(|((idx, oracle_mapping), oracle_type)| {
+                let id: u16 = idx.try_into()?;
+                let oracle_conf = TokenConfig {
+                    token_pair: "".to_string(),
+                    oracle_type: oracle_type.try_into()?,
+                    max_age: None,
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/scope_client.rs:161]: Here we recreate a "config" from the information we can find in the onchain accounts. The accounts don't store a max_age, this is pure bot information (at least for now) to know when to refresh a price. So we set it to None and this will be set to the default max_age when creating the entry from this regenerated config.

File: off_chain/scope-cli/src/scope_client.rs [hunk 7]
```diff
@@ -141,254 +125,215 @@ impl ScopeClient {
...
-                    None
-                }
+            .map(|(id, entry)| {
+                (
+                    *id,
+                    TokenConfig {
+                        token_pair: entry.to_string(),
+                        oracle_mapping: *entry.get_mapping_account(),
+                        oracle_type: entry.get_type(),
+                        max_age: None,
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/scope_client.rs:184]: The only case we use this method is when we download onchain configuration to a configuration file (download command of `scope-cli`). In this case, we have no knowledge of `max_age` as explained above. Even if `entry` is now populated with the `default_max_age` I decided to not put it in the generated config file as it make the edition harder with the hypothesis that most entry will use a default max age that we want to define only once.

I will add a comment to make this clear.

File: off_chain/scope-cli/src/scope_client.rs [hunk 8]
```diff
@@ -141,254 +125,215 @@ impl ScopeClient {
...
-        let old_price = yi_price.value;
-        if new_price != old_price {
-            self.ix_refresh_yi_token_price(yi_idx.try_into()?)?;
-            trace!(
-                "Prices for Yi Token updated successfully at yi_idx {}",
-                yi_idx
-            );
-        } else {
-            trace!("Price for Yi Token has not changed");
+        for (id, ttl) in &prices_ttl {
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/scope_client.rs:255]: They are similar but different enough to make the extraction of a function harder to read and understand from what I tried already.

File: programs/mock_oracles/src/lib.rs [hunk 9]
```diff
@@ -122,7 +143,12 @@ pub mod mock_oracles {
 
         let slot = ctx.accounts.clock.slot;
         price_oracle.valid_slot = slot;
-        msg!("Price {} updated to {} at slot {}", oracle.key, price, slot);
+        msg!(
+            "Pyth price {} updated to {} at slot {}",
```
[the remark this answers is not in the document]
Comment (oeble) [reply programs/mock_oracles/src/lib.rs:147]: For a given token pyth have two accounts that they name "price" and "product". This was until your recent PR a pyth mock program thus the name "price". but we probably can find a better name now yes.

File: programs/scope/src/utils/ctokens.rs (hunk 1 above)
Comment (oeble) [reply programs/scope/src/utils/ctokens.rs]: Yes we use the solend code here.

File: programs/scope/src/utils/mod.rs (hunk 2 above)
Comment (oeble) [reply programs/scope/src/utils/mod.rs:68]: This is okish like that because we trust the admin to provide appropriate accounts there. But as a safety net (not security) I would like some basic checks to avoid a mistake like a bad association between an Oracle type and pubkey so we don't accidentally have (for a short time but still...) a bad account. Maybe we can just check that it deserializes correctly but I think for example that the YiToken is actually a PDA derived from the mint.

File: programs/scope/src/utils/yitoken.rs (hunk 3 above)
Comment (oeble) [reply programs/scope/src/utils/yitoken.rs:30]: Ok but the price is then not the same when we stake and when we unstake (they won't reimburse the fee) so what do we want exactly in scope?
@y2kappa what's your opinion?

File: tests/utils.ts (hunk 4 above)
Comment (oeble) [reply tests/utils.ts:30]: Let's create an issue for a follow-up

File: off_chain/scope-cli/src/config/token_list.rs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+use anyhow::Result;
+use nohash_hasher::IntMap;
+use serde::{Deserialize, Serialize};
+
+use super::token_config::TokenConfig;
+use super::utils::serde_int_map;
+
+/// Format of storage of Scope configuration
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct TokensConfig {
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/config/token_list.rs:12]: renamed to `ScopeConfig`

File: off_chain/scope-cli/src/oracle_helpers/single_account_oracle.rs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+use scope::utils::OracleType;
+use scope::{DatedPrice, Pubkey};
+
+use crate::config::TokenConfig;
+
+use super::{OracleHelper, TokenEntry};
+use anchor_client::solana_sdk::clock;
+
+pub struct SingleAccountOracle {
+    pub label: String,
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/oracle_helpers/single_account_oracle.rs:17]: As discussed, renamed all of them `label`

File: off_chain/scope-cli/src/scope_client.rs [hunk 12]
```diff
@@ -141,254 +125,215 @@ impl ScopeClient {
...
-                            "chunk" = ?chunk,
-                            "first_failed_id" = ?id,
-                            "Refresh of some prices failed"
-                        );
-                    }
-                }
-            } else {
-                trace!("Chunk is too recent, stop");
-                break;
+        for (id, entry) in &self.tokens {
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/scope_client.rs:213]: As discussed, it's getting too verbose

File: off_chain/scope-cli/src/scope_client.rs [hunk 13]
```diff
@@ -141,254 +125,215 @@ impl ScopeClient {
...
+            let age = current_slot.saturating_sub(price_slot);
+            let mut remaining_slots = entry.get_max_age().saturating_sub(age);
+            match entry.need_refresh(price, &self.get_rpc()) {
+                Ok(true) => remaining_slots = 0,
+                Err(e) => error!(
+                    ?e,
+                    "token" = id,
+                    "Error while checking if price need refresh"
+                ),
+                _ => (), // Nothing to do
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/scope_client.rs:306]: Ok for the Ok(false) but as discussed the goal of this match is not to be the complete function return just a possible "alter" of `remaining_slots` so.

File: programs/mock_oracles/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (oeble) [reply programs/mock_oracles/src/lib.rs:147]: as discussed, renamed the the "price" account in the mock_oracle instruction context to `oracle_account`

**Review by andreihrs — COMMENTED** [review]
nice one! I think we need to decide on what we discussed with the cSOL pricing. aside from that, everything looks good to me!

File: tests/utils.ts (hunk 4 above)
Comment (oeble) [reply tests/utils.ts:30]: 15 decimals are the precision of the rate not the number of decimals of the token (this is a completely different thing). I choose a high value here to increase the precision of the computation. Since the rate is close to 1 most of the significative digits are decimals.

File: tests/oracle_utils/switchboard_v1.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  const collateralTokenFeed = new web3.Keypair();
+
+  await mockOracleProgram.rpc.initializeSwitchboardV1(mantissa, scale, {
+    accounts: { oracleAccount: collateralTokenFeed.publicKey, clock: SYSVAR_CLOCK_PUBKEY },
+    signers: [collateralTokenFeed],
+    instructions: [
+      web3.SystemProgram.createAccount({
+        fromPubkey: mockOracleProgram.provider.wallet.publicKey,
+        newAccountPubkey: collateralTokenFeed.publicKey,
+        space: 2500,
```
[the remark this answers is not in the document]
Comment (oeble) [reply tests/oracle_utils/switchboard_v1.ts:16]: This is unchanged from the previous implementation but maybe @gmmour1 can tel.

File: programs/scope/src/utils/mod.rs [hunk 15]
```diff
@@ -21,18 +22,50 @@ pub fn check_context<T>(ctx: &Context<T>) -> ProgramResult {
...
-pub fn get_price(price_type: PriceType, price_acc: &AccountInfo) -> crate::Result<DatedPrice> {
+/// Get the price for a given oracle type
+///
+/// The `base_account` should have been checked against the oracle mapping
+/// If needed the `extra_accounts` will be extracted from the provided iterator and checked
+/// with the data contained in the `base_account`
+pub fn get_price<'a, 'b>(
+    price_type: OracleType,
+    base_account: &AccountInfo,
+    extra_accounts: &mut impl Iterator<Item = &'b AccountInfo<'a>>,
```
[the remark this answers is not in the document]
Comment (oeble) [reply programs/scope/src/utils/mod.rs:42]: Short answer: to please the compiler.

Long answer:
I want this function to accept any kind of iterator that yields references to `AccountInfo`, as such I need to specify the lifetime constraints here so the compiler can check that the yield references are valid when used but more important export correctly the constraint on the lifetime on those references to the caller. The important part here is actually a few lines below:
```rust
where
    'a: 'b,
```
This tells the compiler that the function requires the lifetime `'a` to be longer than `'b`. This seems a logical requirement (since the lifetime `'a` is actually the lifetime of references inside the yields `AccountInfo` we cannot have a reference to a valid `AccountInfo` if the references it contains are not valid anymore) but the compiler doesn't infer everything and here it requires explicit annotations.

Note that for the simpler cases of the lifetime of `base_account` and `extra_accounts`, they are implicit and they are just "must live until the function returns" since the return of this function is not a reference.

File: tests/oracle_utils/switchboard_v1.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (oeble) [reply tests/oracle_utils/switchboard_v1.ts]: Checked is 2500. Constants added.

File: off_chain/scope-cli/src/config/scope_config.rs [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+            TokenConfig {
+                label: "STSOL/USD".to_string(),
+                max_age: None,
+                oracle_mapping: Pubkey::from_str("9LNYQZLJG5DAyeACCTzBFG6H3sDhehP5xtYLdhrZtQkA")
+                    .unwrap(),
+                oracle_type: OracleType::SwitchboardV2,
+            },
+        );
+        token_conf_list.tokens.insert(
+            14, // 4 to test actual holes
```
[the remark this answers is not in the document]
Comment (oeble) [reply off_chain/scope-cli/src/config/scope_config.rs]: Bad copy-paste. Removed.

File: programs/scope/src/utils/ctokens.rs [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        value,
+        exp: DECIMALS.into(),
+    };
+    let dated_price = DatedPrice {
+        price,
+        last_updated_slot: reserve.last_update.slot,
+        _reserved: Default::default(),
+    };
+
+    Ok(dated_price)
```
[the remark this answers is not in the document]
Comment (oeble) [reply programs/scope/src/utils/ctokens.rs:24]: I will implement something on Hubble side.

File: tests/utils.ts (hunk 4 above)
Comment (oeble) [reply tests/utils.ts:30]: Created #34

**Review by y2kappa — APPROVED** [review]

File: configs/devnet/hubble.json [hunk 18]
```diff
@@ -1,122 +1,94 @@
-[
```
[the remark this answers is not in the document]
Comment (oeble) [reply configs/devnet/hubble.json:1]: Scope can manage multiple price feeds and this one is the one used by bubble so...

File: programs/scope/src/utils/yitoken.rs (hunk 3 above)
Comment (oeble) [reply programs/scope/src/utils/yitoken.rs:30]: Pushed to issue #35

File: programs/scope/src/utils/mod.rs (hunk 2 above)
Comment (oeble) [reply programs/scope/src/utils/mod.rs:68]: Follow-up in #36
